argus/db/testrun.py

```diff
@@ -564,7 +565,10 @@ def _heartbeat_entry(self):
             self._log.info("Sending heartbeat...")
             self.heartbeat = time.time()
             self.save()
+        self._log.info("Heartbeat exit")
 
     def shutdown(self):
         self._shutdown_event.set()
+        self._log.info("Waiting for the heartbeat thread to exit...")
+        self._thread.join()
```

Comment:
We can't wait forever

```suggestion
        self._thread.join(timeout=30)
        if self._thread.is_alive():
            self._log.warning("Heartbeat thread was not able to shut down correctly. Stack trace:")
            stack_trace=traceback.extract_stack(sys._current_frames()[self._thread.ident])
              self._log.warning("\n".join([f'#{lineno:3} : {line:50}: {fname}' for fname, lineno, _, line in stack_trace])
```

Reply:
Done